src/classes/com/sun/tdk/jcov/runtime/CollectDetect.java

```diff
@@ -120,54 +121,64 @@ private static ThreadInfo infoForThread(long id) {
                 return ti;
             }
         }
-        // this is a new thread, create a new ThreadInfo
-        synchronized (underConstruction) {
-            // set up a place holder to protect us
-            underConstruction.id = id;
-            underConstruction.next = info[hash];
-            info[hash] = prevInfo = underConstruction;
-
-            // we are now protected, safe to create the real one
-            ti = new ThreadInfo(id); // the new will trigger a track
-            ti.next = underConstruction.next;
-            info[hash] = prevInfo = ti;
+        if (underConstruction != null) {
+            synchronized (underConstruction) {
+                // set up a placeholder to protect us
+                underConstruction.id = id;
+                underConstruction.next = info[hash];
+                info[hash] = underConstruction;
+                prevInfo = underConstruction;
+
+                // we are now protected, safe to create the real one
+                ti = new ThreadInfo(id); // the new will trigger a track
+                ti.next = underConstruction.next;
+                info[hash] = ti;
+                prevInfo = ti;
+            }
         }
         return ti;
     }
 
     public static void hit(int slot) {
-        Thread t = Thread.currentThread();
-        if ( t != null ) {
-            long id = Thread.currentThread().getId();
-            ThreadInfo ti = prevInfo;
+        if (isVMReady()) {
+            Thread t = Thread.currentThread();
+            if (t != null) {
+                long id = Thread.currentThread().getId();
+                ThreadInfo ti = prevInfo;
 
-            if (ti.id != id) {
-                ti = infoForThread(id);
-            }
+                if (ti == null || ti.id != id) {
+                    ti = infoForThread(id);
+                }
 
-            if (ti.enabled()) {
-                Collect.hit(slot);
+                if (ti == null || ti.enabled()) {
+                    Collect.hit(slot);
+                }
             }
         }
     }
 
     public static void hit(int slot, int hash, int fullHash) {
-        Thread t = Thread.currentThread();
-        if( t != null ) {
-            long id = t.getId();
-            ThreadInfo ti = prevInfo;
-
-            if (ti == null || ti.id != id) {
-                ti = infoForThread(id);
-            }
-
-            if (ti.enabled(hash)) {
-                ti.expected = 0;
-                Collect.hit(slot);
-            }
-            if (ti.enabledFull(fullHash)) {
-                ti.expectedFull = 0;
-                Collect.hit(slot);
+        if (isVMReady()) {
+            Thread t = Thread.currentThread();
```

Comment:
Perhaps implement an enhancement later to save into on what coverage is lost because of VM not been ready.